malcolm/modules/ADPandABlocks/parts/kinematicssavupart.py

```diff
@@ -0,0 +1,353 @@
+        # Get the cs port mapping for this PMAC
+        # {scannable: MotorInfo}
+        self.layout_table = context.block_view(self.pmac_mri).layout.value
+        axis_mapping = pmac.util.cs_axis_mapping(
+            context, self.layout_table, axesToMove
+        )
+
+        if self.cs_port is None:
+            # All axes will be in the same cs_port so just use the first
+            for mapping in axis_mapping.values():
+                self.cs_port = mapping.cs_port
+                break
+
+        # Create the mapping of output q variables to axis names
+        for mapping in axis_mapping.values():
+            if mapping.cs_axis in pmac.util.CS_AXIS_NAMES:
+                q_value = pmac.util.CS_AXIS_NAMES.index(mapping.cs_axis) + 1
+                self.q_value_mapping[q_value] = mapping.scannable
+
+        assert "." in self.nxs_full_filename, \
+            "File extension for %r should be supplied" % self.nxs_full_filename
+
+    @add_call_types
+    def post_configure(self, context, part_info):
+        # type: (scanning.hooks.AContext, scanning.hooks.APartInfo) -> None
+
+        self.pos_table = context.block_view(self.panda_mri).positions.value
+
+        # Get the axis number for the inverse kinematics mapped in this cs_port
+        self.axis_numbers = pmac.util.cs_axis_numbers(
+            context, self.layout_table, self.cs_port
+        )
+
+        # Map these in the file
+        dataset_infos = scanning.infos.DatasetProducedInfo.filter_values(
+            part_info
+        )
+        for scannable, axis_num in self.axis_numbers.items():
+            min_i, max_i, value_i = None, None, None
+            for info in dataset_infos:
+                if info.name.startswith(scannable + "."):
+                    if info.type == scanning.infos.DatasetType.POSITION_MIN:
+                        min_i = info
+                    elif info.type == scanning.infos.DatasetType.POSITION_MAX:
+                        max_i = info
+                    elif info.type == scanning.infos.DatasetType.POSITION_VALUE:
+                        value_i = info
+            # Always make sure .value is there
+            assert value_i, "No value dataset for %s" % scannable
+            self.p_vars.append(PVar(
+                path=value_i.path, file=value_i.filename,
+                p_number="p%dmean" % axis_num)
+            )
+            if min_i and max_i:
+                self.p_vars.append(
+                    PVar(
+                        path=min_i.path, file=min_i.filename,
+                        p_number="p%dmin" % axis_num
+                    )
+                )
+                self.p_vars.append(PVar(path=max_i.path, file=max_i.filename,
+                                        p_number="p%dmax" % axis_num))
+            else:
+                self.use_min_max = False
+
+        # Get Forward Kinematics code lines and I,P,M,Q input variables
+        pmac_status_child = context.block_view(self.pmac_mri + ":STATUS")
```

Comment:
Hmm, its a shame we have to hardcode the mri of the status block, but I guess there's no better way really...

Reply:
I did run into several issues whilst testing due to hard-coded mri names not being correct for the given malcolm and clearly I missed this one. Perhaps status_mri_suffix or just status_mri could be passed to the part constructor? (we already pass cs_mri_suffix)

Reply:
yes, probably pass it in the constructor, defaulting to what it is at the moment